### ISSUE TITLE
ONNX export constant folding messes up with shared weight deduplication

### ISSUE THREAD

**Issue by fxmarty** [pytorch/pytorch#108342]
### 🐛 Describe the bug

Hi, it appears that using `do_constant_folding=True` in the ONNX export will undo some weight deduplication. For example, an `nn.Linear` weight will go from

![image](https://github.com/pytorch/pytorch/assets/9808326/5e0a9e87-8c2a-4f23-ae7c-4c4548e95641)

& 

![image](https://github.com/pytorch/pytorch/assets/9808326/752aabad-eeda-44eb-bb36-326c692ba242)

to

![image](https://github.com/pytorch/pytorch/assets/9808326/19c8e6c4-00f5-4973-bf0c-356670348c78)

effectively transposing the weight. Given that the `DeduplicateInitializersByDataPtr` relies on the tensor size, in case the shared weight has a different size (e.g. an embedding weight), the deduplication pass will fail.

It seems to me that the initializer deduplication should happen before the constant folding, and constant folding should be done only for non-shared weights.

WDTY @justinchuby @BowenBao ?

Thank you!

Repro:
```
pip install optimum
optimum-cli export onnx -m bigscience/bloom-560m bloom_onnx --no-post-process
```

and inspect the output with netron

### Versions

Both on 2.0.1 and nightly

**Comment by justinchuby** [comment]
I don't see there's an issue with the suggestion. @bowenbao may have more info

**Comment by BowenBao** [comment]
I agree with the goal in the OP. It seems reasonable for constant folding to detect dup initializers and only fold when all the users of the initializers are foldable and yield same result. Otherwise duplication is introduced. We can probably also introduce heuristics like only process this guard for initializers larger than a certain size, since we don't want to pose this restriction on constant folding small tensors like the ones used in indexing etc.

**Comment by thiagocrepaldi** [comment]
@fxmarty as in previous issues, could you retry this using the latest onnx exporter?

[quick torch.onnx.dynamo_export API tutorial](https://pytorch.org/tutorials/beginner/onnx/export_simple_model_to_onnx_tutorial.html)
